src/Psalm/Internal/Type/TypeExpander.php

```diff
@@ -299,6 +302,56 @@ function ($constant_name) use ($const_name_part): bool {
             return $return_type;
         }
 
+        if ($return_type instanceof TPropertiesOf) {
+            if ($return_type->fq_classlike_name === 'self' && $self_class) {
```

Comment:
Take the opportunity to break this section out into its own method, which will reduce the overall method complexity